python-package/lightgbm/basic.py

```diff
@@ -165,6 +175,39 @@ def list_to_1d_numpy(data, dtype=np.float32, name='list'):
                         "It should be list, numpy 1-D array or pandas Series".format(type(data).__name__, name))
 
 
+def is_numpy_2d_array(data):
+    """Check whether data is a numpy 2-D array."""
+    return isinstance(data, np.ndarray) and len(data.shape) == 2 and data.shape[1] > 1
+
+
+def is_2d_list(data):
+    """Check whether data is a 2-D list."""
+    return isinstance(data, list) and is_1d_list(data[0])
```

Comment:
Should I add a `len(data) > 0` there? I think if someone sends an empty list as init score there's going to be an error somewhere, not sure if this should let it go through though.

Reply:
I think `len(data) > 0` here would be good to have right before passing `dat[0]` to `is_1d_list()`.

Reply:
Just want to note something I learned recently, that `scipy` sparse matrices don't support `len()`.

```python
from scipy.sparse import csr_matrix
X = csr_matrix((1,1))
len(X)
```

> Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jlamb/miniconda3/lib/python3.7/site-packages/scipy/sparse/base.py", line 291, in __len__
    raise TypeError("sparse matrix length is ambiguous; use getnnz()"
TypeError: sparse matrix length is ambiguous; use getnnz() or shape[0]

I don't think it would be a problem in this specific case since the check comes after `is isinstance(data, list) and` so `len(data)` wouldn't be evaluated in that statement if `data` is not a list, but just wanted to note it in case other refactorings here result in calling `len(data)` when the type of `data` hasn't been checked yet.